compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -421,13 +427,58 @@ impl server::TokenStream for Rustc<'_> {
         parse_stream_from_source_str(
             FileName::proc_macro_source_code(src),
             src.to_string(),
-            self.sess,
+            self.sess(),
             Some(self.call_site),
         )
     }
     fn to_string(&mut self, stream: &Self::TokenStream) -> String {
         pprust::tts_to_string(stream)
     }
+    fn expand_literal(&mut self, stream: &Self::TokenStream) -> Result<Self::Literal, ()> {
+        // Parse the expression from our tokenstream.
+        let expr = {
+            let mut p = rustc_parse::stream_to_parser(
+                self.sess(),
+                stream.clone(),
+                Some("proc_macro expand literal"),
+            );
+            let expr = p.parse_expr().map_err(|_| ())?;
+            if p.token != token::Eof {
+                return Err(());
+            }
+            expr
+        };
+
+        // Perform eager expansion on the expression.
+        let expr = self
+            .ecx
+            .expander()
+            .fully_expand_fragment(crate::expand::AstFragment::Expr(expr))
+            .make_expr();
+
+        // Extract the literal value from the expression.
+        let lit = match &expr.kind {
+            ast::ExprKind::Lit(l) => l.kind.to_lit_token(),
+            ast::ExprKind::Unary(ast::UnOp::Neg, e) => match &e.kind {
+                ast::ExprKind::Lit(l) => match l.kind.to_lit_token() {
+                    t @ token::Lit { kind: token::Integer | token::Float, .. } => {
+                        token::Lit::new(t.kind, Symbol::intern(&format!("-{}", t.symbol)), t.suffix)
+                    }
+                    _ => return Err(()),
+                },
+                _ => return Err(()),
+            },
+            _ => return Err(()),
+        };
+
+        // Filter to valid literal kinds to return from this method, and
+        // sanitize the returned span to ensure the calling proc_macro can't use
+        // them.
+        match lit.kind {
+            token::Bool | token::Err => Err(()),
+            _ => Ok(Literal { lit, span: expr.span.with_ctxt(self.call_site.ctxt()) }),
```

Comment:
This sanitizing doesn't make sense to me and I'm honestly don't understand the amount of FUD related to this in https://github.com/rust-lang/rust/issues/67984 and the relevant RFC thread.
We should preserve the hygienic context here, and specify what context is assigned to literals returned by built-in macros - call-site (most likely), def-site, or explicitly unspecified. In the same way like we specify that for identifiers.

Reply:
My concern with preserving it was mostly theoretical, and focused on a concern around what might happen if someone invoked a std macro which has `#[allow_internal_unstable(...)]` on it and exposes a literal. I'm OK with dropping the sanitizing if we're confident it won't be an issue, or want to leave figuring it out for later.